### PR TITLE
MEN-4482, MEN-4418: implement file upload and download end-points

### PR DESCRIPTION
**Pull request by tranchitella**
Changelog: none

Signed-off-by: Fabio Tranchitella <fabio.tranchitella@northern.tech>

Depends on: https://github.com/mendersoftware/mender-connect/pull/17

### CODE REVIEW

File: api/http/management_filetransfer.go [hunk 1]
```diff
@@ -50,11 +50,11 @@ type fileTransferParams struct {
...
-	hdrMenderFileTransferPath = "X-Mender-Path"
-	hdrMenderFileTransferUID  = "X-Mender-UID"
-	hdrMenderFileTransferGID  = "X-Mender-GID"
-	hdrMenderFileTransferMode = "X-Mender-Mode"
-	hdrMenderFileTransferSize = "X-Mender-Size"
+	hdrMenderFileTransferPath = "X-Men-File-Path"
+	hdrMenderFileTransferUID  = "X-Men-File-UID"
+	hdrMenderFileTransferGID  = "X-Men-File-GID"
+	hdrMenderFileTransferMode = "X-Men-File-Mode"
+	hdrMenderFileTransferSize = "X-Men-File-Size"
```
Comment (merlin-northern) [inline api/http/management_filetransfer.go:57]: headers are case-insensitive I believe in http, but in the RBAC case we use `X-MEN-`

File: docs/management_api.yml [hunk 2]
```diff
@@ -61,6 +61,40 @@ paths:
                 request_id: "eed14d55-d996-42cd-8248-e806663810aa"
         500:
           $ref: '#/components/responses/InternalServerError'
+  /devices/{id}/check-update:
```
Comment (merlin-northern) [inline docs/management_api.yml:64]: wait how is that related to the File Transfer?

File: docs/management_api.yml [hunk 3]
```diff
@@ -120,22 +154,59 @@ paths:
...
-      operationId: Check Update
-      summary: Trigger check-update for the Mender client running on the device
+      operationId: Download
+      summary: Download a file from the device
       parameters:
         - in: path
           name: id
           required: true
           schema:
             type: string
```
Comment (merlin-northern) [inline docs/management_api.yml:168]: how about:
```
          schema:
            type: string
            format: uuid
```

File: docs/management_api.yml [hunk 4]
```diff
@@ -120,22 +154,59 @@ paths:
...
           description: ID for the target device.
```
Comment (merlin-northern) [inline docs/management_api.yml:169]: minor: "target device" somehow I would change to "device" or to "device to get the file from" (feel free to ignore)

File: docs/management_api.yml [hunk 5]
```diff
@@ -120,22 +154,59 @@ paths:
...
+      requestBody:
+        content:
+          'application/json':
+            schema:
+              type: object
+              properties:
+                path:
+                  type: string
+                  description: Path of the file on the target device.
```
Comment (merlin-northern) [inline docs/management_api.yml:178]: technically speaking true, but it is "a path to the file to download", feel free to disregard if you do not like it.

File: docs/management_api.yml [hunk 6]
```diff
@@ -120,22 +154,59 @@ paths:
...
+                path:
+                  type: string
+                  description: Path of the file on the target device.          
+              required:
+                - path
       responses:
-        202:
-          description: Accepted
+        200:
+          description: The content of the file will be returned as attachment
```
Comment (merlin-northern) [inline docs/management_api.yml:183]: I find "attachment" a bit strange here.

File: docs/management_api.yml [hunk 7]
```diff
@@ -255,6 +326,45 @@ paths:
         500:
           $ref: '#/components/responses/InternalServerError'
 
+  /devices/{id}/upload:
+    post:
```
Comment (merlin-northern) [inline docs/management_api.yml:330]: idle question, do we consider overwriting a file with PUT? what do we do in general when a file exists?

File: docs/management_api.yml [hunk 8]
```diff
@@ -255,6 +326,45 @@ paths:
...
+      tags:
+        - ManagementAPI
+      operationId: Upload
+      summary: Upload a file to the device
+      parameters:
+        - in: path
+          name: id
+          required: true
+          schema:
+            type: string
```
Comment (merlin-northern) [inline docs/management_api.yml:340]: how about:
```
          schema:
            type: string
            format: uuid
```

File: docs/management_api.yml [hunk 9]
```diff
@@ -120,22 +154,59 @@ paths:
...
+              properties:
+                path:
+                  type: string
+                  description: Path of the file on the target device.          
+              required:
+                - path
       responses:
-        202:
-          description: Accepted
+        200:
```
Comment (merlin-northern) [inline docs/management_api.yml:182]: the errors are not part of the diff (nice way to disallow comments! ;-)) but does 404 mean file not found or device not found?

File: docs/management_api.yml [hunk 10]
```diff
@@ -298,6 +408,26 @@ components:
...
 
+    FileUpload:
+      type: object
+      properties:
+        path:
+          type: string
+          description: The path of the file on the device
+        uid:
+          type: integer
+          description: The numerical UID of the file on the device
```
Comment (merlin-northern) [inline docs/management_api.yml:419]: I would explicitly note that this UID is the local user id on the device and has no meaning outside it. Also for these reasons I have doubts if it s worthwhile to return the ids, for the UI it would be of some value to get the username and the name of the group, if we want to present it. otherwise we need another protocol message GETUID/GETGID to translate.

File: api/http/management_filetransfer.go [hunk 11]
```diff
@@ -516,7 +516,7 @@ func (h ManagementController) UploadFile(c *gin.Context) {
 		request.GID = &nGID
 	}
 	if mode := c.Request.Form.Get(fieldUploadMode); mode != "" {
-		v, _ := strconv.Atoi(mode)
+		v, _ := strconv.ParseInt(mode, 8, 64)
```
Comment (merlin-northern) [inline api/http/management_filetransfer.go:519]: are we sure we are ignoring error here?

File: docs/management_api.yml (hunk 2 above)
Comment (tranchitella) [reply docs/management_api.yml:64]: I simply moved it to have it in alphabetical order

File: api/http/management_filetransfer.go (hunk 11 above)
Comment (tranchitella) [reply api/http/management_filetransfer.go:519]: yes, default is zero which is the null value of int

File: docs/management_api.yml (hunk 7 above)
Comment (tranchitella) [reply docs/management_api.yml:330]: the `/upload` endpoint overwrites the files. I wasn't sure about PUT vs. POST, but as we accept a multipart/form-data payload, I thought POST is more appropriate. @alfrunes what do you think?

File: docs/management_api.yml (hunk 9 above)
Comment (tranchitella) [reply docs/management_api.yml:182]: Device not found. File not found is Bad Request.

File: docs/management_api.yml (hunk 10 above)
Comment (tranchitella) [reply docs/management_api.yml:425]: I don't think we should increase the complexity and implement getUID/getGID or similar. The issue is that you won't be able to set non-numerical owners to the local file anyway, unless you have them in your local system. @alfrunes's idea was to stick to the numerical IDs because at least they are portable, and this is what you do when you us tar with `--numeric-owner`. What shall we do here?

Comment (merlin-northern) [reply docs/management_api.yml:425]: when you pass a username and a group name to the mender-connect in the message which creates a file, then locally you resolve to uid and gid to set the owner/group on the local file. passing and storing the ids of a local system outside of it seems a bit strange.
and it is no accident that tar has `--numeric-owner` it is exactly this use case.

Comment (merlin-northern) [reply docs/management_api.yml:425]: but if you both say it is ok, then I will not argue too much.

File: docs/management_api.yml (hunk 7 above)
Comment (alfrunes) [reply docs/management_api.yml:335]: According to the [specification](https://tools.ietf.org/html/rfc7231#section-4.3.4) it sounds more correct to use `PUT` here actually.

**Review by alfrunes — CHANGES_REQUESTED** [review]
I found a couple of issues; most importantly the timer allocations inside the upload/download loops and the mode description not matching the handler for upload.
The rest are primarily suggestions.

File: docs/management_api.yml (hunk 10 above)
Comment (alfrunes) [reply docs/management_api.yml:425]: I tried to keep the implementation at the filesystem level, without involving OS specific user management. I think you can use the same argument with OS user/group names; they don't have any meaning outside the device's OS, and may have different numerical UIDs/GIDs across different devices.
I'm open to change it if it's a bad idea, but these were my thoughts while developing this.

File: api/http/management_filetransfer.go [hunk 12]
```diff
@@ -0,0 +1,744 @@
...
+	defer ticker.Stop()
+
+	// handle messages from the device
+	latestMessage := time.Now()
+	latestOffset := int64(0)
+	numberOfChunks := 0
+	for {
+		ctx, cancel := context.WithDeadline(context.Background(),
+			latestMessage.Add(fileTransferTimeout))
+		defer cancel()
```
[the remark this answers is not in the document]
Comment (alfrunes) [reply api/http/management_filetransfer.go:290]: I did a similar thing in mender-connect, and the qemu device ran out of memory pretty quick when transferring a larger file.

File: docs/management_api.yml (hunk 10 above)
Comment (tranchitella) [reply docs/management_api.yml:425]: we won't change this at this stage

File: api/http/management_filetransfer.go [hunk 13]
```diff
@@ -0,0 +1,744 @@
...
+						latestOffset); err != nil {
+						return
+					}
+					numberOfChunks = 0
+				}
+			}
+
+		// send a Ping message to keep the session alive
+		case <-ticker.C:
+			h.publishFileTransferPing(params.SessionID, deviceTopic)
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply api/http/management_filetransfer.go:371]: Does it hurt if it stays?

File: api/http/management_filetransfer.go [hunk 14]
```diff
@@ -0,0 +1,744 @@
...
+		})
+		return
+	}
+
+	h.uploadFileResponse(c, params, request)
+}
+
+func (h ManagementController) fileTransferAllowed(c *gin.Context, tenantID string,
+	deviceID string) (bool, error) {
+	if len(c.Request.Header.Get(model.RBACHeaderRemoteTerminalGroups)) == 0 {
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply api/http/management_filetransfer.go:738]: Yes, we are not introducing a new role/permission, and reusing the remote terminal groups header.
Not super clean, but I don't know how to handle it without messing it up with RBAC.

**Comment by tranchitella** [comment]
@alfrunes ready for another round of review, fixes are here: https://github.com/mendersoftware/deviceconnect/pull/51/commits/83356f45de5aedd11e54ebfa3d721037d50acf44

File: api/http/management_filetransfer.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (alfrunes) [reply api/http/management_filetransfer.go:369]: Not at all, but we also need to deal with inbound ping/pong messages as well.

File: api/http/management_filetransfer.go [hunk 15]
```diff
@@ -0,0 +1,736 @@
...
+
+			// process incoming messages from the device by type
+			switch msg.Header.MsgType {
+
+			// error message, stop here
+			case wsft.MessageTypeError:
+				errorMsg := msgBody.(*wsft.Error)
+				errorChan <- errors.New(*errorMsg.Error)
+				return
+
```
Comment (alfrunes) [inline api/http/management_filetransfer.go:446]:
```suggestion

			case ws.MessageTypePing:
				// send pong
			case ws.MessageTypePong:
				// update timeout
```

File: api/http/management_filetransfer.go [hunk 16]
```diff
@@ -0,0 +1,736 @@
...
+		}
+		return msg, msgBody, nil
+	case wsft.MessageTypeFileInfo:
+		msgBody := &wsft.FileInfo{}
+		err := msgpack.Unmarshal(msg.Body, msgBody)
+		if err != nil {
+			return nil, nil, errors.Wrap(err, errFileTranserUnmarshalling.Error())
+		}
+		return msg, msgBody, nil
+	case wsft.MessageTypeACK, wsft.MessageTypeChunk:
```
Comment (alfrunes) [inline api/http/management_filetransfer.go:203]:
```suggestion
	case wsft.MessageTypeACK, wsft.MessageTypeChunk, ws.MessageTypePing, ws.MessageTypePong:
```

File: api/http/management_filetransfer.go [hunk 17]
```diff
@@ -0,0 +1,736 @@
...
+					if err := h.publishFileTransferProtoMessage(
+						params.SessionID, params.UserID, deviceTopic,
+						wsft.MessageTypeACK, nil,
+						latestOffset); err != nil {
+						return
+					}
+					numberOfChunks = 0
+				}
+			}
+
```
Comment (alfrunes) [inline api/http/management_filetransfer.go:366]:
```suggestion

			case ws.MessageTypePing:
				// send pong
				if err := h.publishFileTransferProtoMessage(
						params.SessionID, params.UserID, deviceTopic,
						ws.MessageTypePong, nil,
						-1); err != nil {
						return
					}
				fallthrough
			case ws.MessageTypePong:
				ticker.Reset(fileTransferPingInterval)
```
```

**Review by alfrunes — COMMENTED** [review]
We also need to handle inbound ping/pong messages.

File: api/http/management_filetransfer.go [hunk 18]
```diff
@@ -202,9 +200,11 @@ func (h ManagementController) decodeFileTransferProtoMessage(data []byte) (*ws.P
 			return nil, nil, errors.Wrap(err, errFileTranserUnmarshalling.Error())
 		}
 		return msg, msgBody, nil
+	case wsft.MessageTypeACK, wsft.MessageTypeChunk:
+		return msg, nil, nil
 	}
 
-	return msg, nil, nil
+	return nil, nil, errors.Errorf("unexpected message type '%s'", msg.Header.MsgType)
```
[the remark this answers is not in the document]
Comment (tranchitella) [reply api/http/management_filetransfer.go:207]: uhm, no... they have their own return, don't they?

File: api/http/management_filetransfer.go (hunk 17 above)
Comment (tranchitella) [reply api/http/management_filetransfer.go:366]: I didn't know the existence of the `fallthrough` keyword!

File: api/http/management_filetransfer.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (alfrunes) [reply api/http/management_filetransfer.go:207]: ohh... Sorry, I missed the final return statements :facepalm:

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline

**Comment by tranchitella** [comment]
Failure in `test_demo_artifact` is spurious

File: api/http/management_filetransfer.go [hunk 19]
```diff
@@ -0,0 +1,760 @@
...
+				return err
+			}
+			*numberOfChunks = 0
+		}
+
+	case ws.MessageTypePing:
+		if err := h.publishFileTransferProtoMessage(
+			params.SessionID, params.UserID, deviceTopic,
+			ws.MessageTypePong, nil,
+			-1); err != nil {
```
Comment (alfrunes) [inline api/http/management_filetransfer.go:391]: This message has wrong protocol ID (`ws.ProtoTypeFIleTransfer` - should have `ws.ProtoTypeSessionControl`) and you'll get an error message in response.

File: api/http/management_filetransfer.go (hunk 15 above)
Comment (alfrunes) [reply api/http/management_filetransfer.go:470]: I'm not seeing ping/pong being handled for upload.

File: api/http/management_filetransfer.go [hunk 20]
```diff
@@ -0,0 +1,760 @@
...
+
+	return nil, nil, errors.Errorf("unexpected message type '%s'", msg.Header.MsgType)
+}
+
+func writeHeaders(c *gin.Context, fileInfo *wsft.FileInfo) {
+	c.Writer.WriteHeader(http.StatusOK)
+	c.Writer.Header().Add(hdrContentType, "application/octet-stream")
+	if fileInfo.Path != nil {
+		filename := path.Base(*fileInfo.Path)
+		c.Writer.Header().Add(hdrContentDisposition, "attachment; filname=\""+filename+"\"")
```
Comment (alfrunes) [inline api/http/management_filetransfer.go:215]:
```suggestion
		c.Writer.Header().Add(hdrContentDisposition, "attachment; filename=\""+filename+"\"")
```

File: api/http/management_filetransfer.go (hunk 19 above)
Comment (tranchitella) [reply api/http/management_filetransfer.go:394]: I don't understand the "and you'll get an error message in response".

File: api/http/management_filetransfer.go (hunk 15 above)
Comment (tranchitella) [reply api/http/management_filetransfer.go:474]: I handled the ping messages, generating pong. I'm not sure it makes sense to do the "pings" when uploading, as we are already supposed to send a lot of messages and wait for acks. What do you think?

File: api/http/management_filetransfer.go (hunk 19 above)
Comment (alfrunes) [reply api/http/management_filetransfer.go:395]: Ohh, sorry. Reading it again I see I'm very imprecise; what I meant is since the message has wrong `proto` header, it will be routed to the handler for file transfer messages which in turn will return an error saying it don't understand the message type.

File: api/http/management_filetransfer.go (hunk 15 above)
Comment (alfrunes) [reply api/http/management_filetransfer.go:475]: I'm pretty sure don't need it for either upload nor download since there is a constant flow of messages in both directions. So this is good enough for me.

**Review by alfrunes — APPROVED** [review]

**Comment by tranchitella** [comment]
@mender-test-bot start pipeline

**Comment by tranchitella** [comment]
Failures are spurious:
- `tests/test_demo_artifact.py::TestDemoArtifactEnterprise::test_demo_artifact`
- `tests/test_mtls.py::TestClientMTLSEnterprise::test_mtls_enterprise_hsm[rsa]`
